### PR TITLE
[Data] Fix legacy do_write

### PR DESCRIPTION
**Pull request by matthew29tang**
## Why are these changes needed?

With the new `write` added (from #32015 and #32440), Ray Data intends to support both the `write` and `do_write` functions for now. The check currently uses the `hasattr()` function to ensure the datasource object has a `write` method before using it.

However, this is insufficient for a custom datasource that inherits from `Datasource` since `Datasource` has the `write` method implemented. If the custom datasource only has `do_write` implemented, `hasattr(datasource, "write")` will return True since `hasattr()` will detect methods via inheritance. 

The solution is to check if the `write` method was overwritten from `Datasource.write`. Any class that has not implemented `write` will have the equality check return True

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by jianoaix — COMMENTED** [review]
Thanks for fixing! Could we add a test for this case?

**Comment by matthew29tang** [comment]
How would you recommend adding a test for this? I don't see existing test cases for CustomDatasource to base off of.

**Comment by jianoaix** [comment]
You may extend the `Datasource` and implement a TestDatasource, similar to https://github.com/ray-project/ray/blob/master/python/ray/data/datasource/datasource.py#L318

**Review by jianoaix — APPROVED** [review]

**Comment by ericl** [comment]
Dataset tests seem to be failing (@jianoaix )

**Comment by matthew29tang** [comment]
Apologies, I will investigate. 

Quick sanity check, how do you run the data tests locally? I am able to run basic tests with `python -m pytest -v python/ray/tests/test_mini.py` with the working directory in the top level ray folder. But the data tests seem to not run with `python -m pytest -v python/ray/data/tests/test_dataset.py` (`ModuleNotFoundError: No module named 'ray.data.tests'`)

**Comment by jianoaix** [comment]
> Apologies, I will investigate.
> 
> Quick sanity check, how do you run the data tests locally? I am able to run basic tests with `python -m pytest -v python/ray/tests/test_mini.py` with the working directory in the top level ray folder. But the data tests seem to not run with `python -m pytest -v python/ray/data/tests/test_dataset.py` (`ModuleNotFoundError: No module named 'ray.data.tests'`)

Yes, that's how we run tests locally. Can you try with `python python/ray/data/tests/test_dataset.py'?

**Comment by matthew29tang** [comment]
Running that in the same directory (top level ray) gives the same error:

```
Traceback (most recent call last):
  File "python/ray/data/tests/test_dataset.py", line 39, in <module>
    from ray.data.tests.conftest import *  # noqa
ModuleNotFoundError: No module named 'ray.data.tests'
```

**Comment by jianoaix** [comment]
> Running that in the same directory (top level ray) gives the same error:
> 
> ```
> Traceback (most recent call last):
>   File "python/ray/data/tests/test_dataset.py", line 39, in <module>
>     from ray.data.tests.conftest import *  # noqa
> ModuleNotFoundError: No module named 'ray.data.tests'
> ```

The test failed is actually this one: https://buildkite.com/ray-project/oss-ci-build-pr/builds/12645

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -2693,7 +2693,7 @@ def write_datasource(
                 soft=False,
             )
 
-        if hasattr(datasource, "write"):
+        if "write" in datasource.__dict__:
```
Comment (jianoaix) [inline python/ray/data/dataset.py:2696]: This doesn't look right. The `datasource.__dict__` will be an empty dict. You may want to check `type(datasource).__dict__`

Comment (matthew29tang) [reply python/ray/data/dataset.py]: good catch

**Comment by matthew29tang** [comment]
To my understanding, the current issue is that various front-facing Datasources do not implement `write`, but they inherit `write` from a parent (ex. `FileBasedDatasource`). This is likely why the initial author used `hasattr()` to check with inheritance. 

The original problem still stands that `hasattr()` is vacuously True since anything that extends `Datasource` will thus have a `write` method (but raise `NotImplementedError`).

Still investigating a solution to this.

**Comment by jianoaix** [comment]
> To my understanding, the current issue is that various front-facing Datasources do not implement `write`, but they inherit `write` from a parent (ex. `FileBasedDatasource`). This is likely why the initial author used `hasattr()` to check with inheritance.
> 
> The original problem still stands that `hasattr()` is vacuously True since anything that extends `Datasource` will thus have a `write` method (but raise `NotImplementedError`).
> 
> Still investigating a solution to this.

Yeah, with the fix this PR, those subclasses of `FileBasedDatasource` will still run the legacy `do_write()` path, although the `write()` has been implemented in `FileBasedDatasource`.

**Comment by matthew29tang** [comment]
Just pushed a new version which checks if the `write` method was overwritten from `Datasource.write`. Any class that has not implemented `write` will have the equality check return True

**Comment by jianoaix** [comment]
> Just pushed a new version which checks if the `write` method was overwritten from `Datasource.write`. Any class that has not implemented `write` will have the equality check return True

That LGTM!

**Comment by jianoaix** [comment]
It looks tests passing now. @matthew29tang Can you update the PR description based on current state?

**Comment by matthew29tang** [comment]
Just edited the PR description.

**Comment by ericl** [comment]
Merged, thanks!

**Pull request merged**